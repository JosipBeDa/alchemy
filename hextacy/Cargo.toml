[package]
authors = ["biblius <crazycompanyxxl@gmail.com>"]
description = "Library containing various utilities to aid server development with hexagonal architecture"
edition = "2021"
homepage = "https://github.com/biblius/hextacy"
keywords = ["backend", "hextacy", "hextc", "xtc"]
license = "MIT"
name = "hextacy"
readme = "../README.md"
version = "0.1.2"

[lib]
name = "hextacy"
path = "src/lib.rs"

[dependencies]
actix = { version = "0.13.0", optional = true }
actix-web = "4"
actix-web-actors = { version = "4.1.0", optional = true }
async-trait = "0.1.42"
bcrypt = "0.14.0"
chrono = { version = "0.4", features = ["serde"] }
data-encoding = "2.3.2"
diesel = { version = "2.0.1", features = [
  "chrono",
  "postgres",
  "r2d2",
  "serde_json",
], optional = true }
dotenv = "0.15.0"
env_logger = "0.10.0"
hextacy_derive = { path = "../hextacy_derive" }
hmac = "0.12.1"
jsonwebtoken = "8.1.1"
lettre = { version = "0.10.3", optional = true }
log4rs = "1.2.0"
mongodb = { version = "2.3.1", features = ["tokio-runtime"], optional = true }
parking_lot = "0.12.1"
r2d2 = { version = "0.8.10", optional = true }
rand = "0.8.5"
redis = { version = "0.22.3", features = ["r2d2"], optional = true }
rsa = { version = "0.8.2", features = ["pem"] }
sea-orm = { version = "0.11.2", features = [
  "macros",
  "runtime-actix-native-tls",
  "sqlx-postgres",
], optional = true }
serde = { version = "1.0.137", features = ["derive"] }
serde_json = "1.0.81"
sha2 = "0.10.6"
thiserror = "1.0.37"
thotp = "0.1.11"
tokio = "1.24.2"
tracing = "0.1.35"
tracing-test = "0.2.3"
uuid = { version = "1.1.2", features = ["v4"] }


[features]
default = ["full"]
full = ["db", "email", "ws"]

db = ["dep:diesel", "dep:mongodb", "dep:r2d2", "dep:redis", "dep:sea-orm"]
ws = ["dep:actix", "dep:actix-web-actors"]

email = ["dep:lettre"]
mongo = ["dep:mongodb"]
postgres-diesel = ["dep:diesel"]
postgres-seaorm = ["dep:sea-orm"]
redis = ["dep:r2d2", "dep:redis"]
