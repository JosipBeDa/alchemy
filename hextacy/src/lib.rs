/// Utility traits for connecting to the cache. Currently supports only redis.
#[cfg(any(feature = "db", feature = "full", feature = "redis"))]
pub mod cache;

/// Traits and macros for implementing drivers and connections.
#[cfg(any(
    feature = "db",
    feature = "full",
    feature = "diesel",
    feature = "mongo"
))]
pub mod db;

/// Drivers for connecting to the database, cache and smtp.
pub mod drivers;

/// Cryptographic utilities
pub mod crypto;

/// Utilities for getting and setting stuff from the env.
pub mod env;

/// A logger that can be set up to use stdout or a file.
pub mod logger;

/// Utilities for time related stuff.
pub mod time;

/// Utilities for web related stuff. Contains a WS and broker implementation
/// as well as some HTTP helpers.
pub mod web;

/// Derive macros for quick implementations of generic repository traits.
pub mod derive {
    pub use hextacy_derive::Repository;
}
