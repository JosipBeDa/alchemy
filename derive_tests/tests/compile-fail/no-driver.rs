use hextacy::{
    derive::Repository,
    drivers::{
        db::{DBConnect, Driver},
        DriverError,
    },
};

trait UserRepository<C> {}
struct Conn;
struct SomeDriver;

#[async_trait::async_trait]
impl DBConnect for SomeDriver {
    type Connection = Conn;
    async fn connect(&self) -> Result<Self::Connection, DriverError> {
        Ok(Conn)
    }
}

#[derive(Repository)]
struct ServiceRepo<D, C, User>
where
    D: DBConnect<Connection = C>,
{
    driver: Driver<D, C>,
    _user: User,
}

fn main() {}
