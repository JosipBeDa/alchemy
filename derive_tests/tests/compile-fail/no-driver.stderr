error: [Acid]Repository needs at least one driver field

         = help: Annotate the Driver field with one of the available drivers: diesel, mongo, seaorm

  --> tests/compile-fail/no-driver.rs:22:8
   |
22 | struct ServiceRepo<D, C, User>
   |        ^^^^^^^^^^^
